manager/controlapi/secret.go

```diff
@@ -31,6 +32,45 @@ func secretFromSecretSpec(spec *api.SecretSpec) *api.Secret {
 	}
 }
 
+// checkSecretInUse does a best effort to find if the passed in secret is in
+// use by any services
+func (s *Server) checkSecretInUse(secretID string) error {
+	var (
+		services []*api.Service
+		err      error
+	)
+
+	s.store.View(func(tx store.ReadTx) {
+		services, err = store.FindServices(tx, store.All)
+	})
+	if err != nil {
+		return err
+	}
+
+	var secretName string
+	servicesUsingSecret := make(map[string]struct{})
+	for _, service := range services {
+		if service.Spec.Task.GetContainer() != nil {
+			for _, secretRef := range service.Spec.Task.GetContainer().Secrets {
+				if secretRef.SecretID == secretID {
+					servicesUsingSecret[service.ID] = struct{}{}
+					secretName = secretRef.SecretName
+				}
+			}
+		}
+	}
+	if len(servicesUsingSecret) > 0 {
+		services := make([]string, 0, len(servicesUsingSecret))
+		for k := range servicesUsingSecret {
+			services = append(services, k)
+		}
+
+		return grpc.Errorf(codes.DataLoss, "secret '%s' is in use by the following service: %v", secretName, strings.Join(services, ", "))
```

Comment:
Hmm, I don't think `DataLoss` is correct.

> `DataLoss indicates unrecoverable data loss or corruption.`

Also, what I had in mind here was to say "service" if `len(services) == 1` and "services" otherwise.